File: problems/happy_number/HappyNumber.java
package problems.happy_number;

public class HappyNumber {
    private int squareSum(int n) {
        int result = 0;
        while (n > 0) {
            int digit = n % 10;
            result += digit * digit;
            n = n / 10;
        }
        return result;
    }

    public boolean isHappy(int n) {
        int fast = n;
        int slow = n;

        while (true) {
            slow = squareSum(slow);
            fast = squareSum(squareSum(fast));
            if (slow == 1) {
                return true;
            }
            if (slow == fast)
                return false;
        }
    }
}

File: problems/happy_number/happy_number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def square_sum(n):
            result = 0
            while n > 0:
                digit = n % 10
                result += digit * digit
                n = n // 10
            return result

        slow = n
        fast = n

        while True:
            slow = square_sum(slow)
            fast = square_sum(square_sum(fast))

            if slow == 1:
                return True
            if slow == fast:
                return False
